### PR TITLE
ci: prepare the release workflow

### PR DESCRIPTION
**Pull request by missedone**

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -28,3 +28,15 @@ jobs:
...
+
+      - name: Publish snapshots
+        uses: samuelmeuli/action-maven-publish@v1
+        # publish snapshot only when push to master branch
+        if: github.event_name == 'push'
+        with:
+          gpg_private_key: ${{ secrets.gpg_private_key }}
+          gpg_passphrase: ${{ secrets.gpg_passphrase }}
+          nexus_username: ${{ secrets.nexus_username }}
+          nexus_password: ${{ secrets.nexus_password }}
```
Comment (rr-nick-tan) [inline .github/workflows/ci.yml:40]: @cbeust could you create the above four secrets at the testng-team github org level?

**Comment by rr-nick-tan** [comment]
fixes #1  , #81

**Comment by juherr** [comment]
@krmahadevan could you try with your credentials?

**Comment by krmahadevan** [comment]
> @krmahadevan could you try with your credentials?

@juherr - I didn't get you. What do you want me to try with my credentials ?

**Comment by juherr** [comment]
@krmahadevan add secrets in order to publish artifacts. I think your credentials will work.

**Comment by krmahadevan** [comment]
@juherr - I am searching for my credentials. Not sure where I have saved them after I lost all data on my Hard drive. Please give me sometime on this.

**Comment by krmahadevan** [comment]
@juherr - I managed to finally find the credentials and I have added them to the repo settings

image

**Comment by juherr** [comment]
@missedone could you check if it is working?

**Comment by missedone** [comment]
@krmahadevan thanks for updating the secrets, however the snapshot publishing failed:
https://github.com/testng-team/testng-remote/runs/6045933404?check_suite_focus=true

image

could you double check set the gpg secrets with the instruction here https://github.com/samuelmeuli/action-maven-publish#secrets

**Comment by krmahadevan** [comment]
@missedone - I have setup the keys once again, just to rule out any manual errors and it still throws the same error.

I looked at the latest raw logs https://pipelines.actions.githubusercontent.com/serviceHosts/14024220-b483-4fbd-97de-92f0294e671b/_apis/pipelines/1/runs/52/signedlogcontent/2?urlExpires=2022-04-22T04%3A32%3A13.4714860Z&urlSigningMethod=HMACV1&urlSignature=EJWacww9NlYboBLWpI61N6ebEi7fbgtOBhZHTpbPOSM%3D


I see the below lines

```
2022-04-16T06:16:58.5415754Z   JAVA_HOME: /opt/hostedtoolcache/Java_Adopt_jdk/11.0.11-9/x64
2022-04-16T06:16:58.5416255Z ##[endgroup]
2022-04-16T06:16:58.5785029Z Importing GPG key…
2022-04-16T06:16:58.5864817Z gpg: directory '/home/runner/.gnupg' created
2022-04-16T06:16:58.5868159Z gpg: keybox '/home/runner/.gnupg/pubring.kbx' created
2022-04-16T06:16:58.5884350Z gpg: /home/runner/.gnupg/trustdb.gpg: trustdb created
2022-04-16T06:16:58.5885782Z gpg: key 0F13D5631D6AF36D: public key "Krishnan Mahadevan (***-key) <krishnan.mahadevan1978@gmail.com>" imported
2022-04-16T06:16:58.5970714Z gpg: Total number processed: 1
2022-04-16T06:16:58.5971416Z gpg:               imported: 1
```

Doesn't this mean that the keys were imported?

**Comment by juherr** [comment]
Gpg should be used during the build step in order to sign the jar file.
It could explain the issue.

**Comment by rr-nick-tan** [comment]
@krmahadevan since i don't have the access of the keys for further diagnosing.
could you try with this command being used by the GHA step:
```bash
 mvn clean deploy --batch-mode --activate-profiles deploy --settings /home/runner/work/_actions/samuelmeuli/action-maven-publish/v1/settings.xml 
```

NOTE: you have to create a temporary settings.xml with the gpg configured there, remember to use the actual path of settings.xml on your local
Thanks

**Comment by krmahadevan** [comment]
@rr-nick-tan - Can you please let me know if there's a template `settings.xml` that I can refer to for creating this file locally? I will try this out later today or over the week and post back the results.

**Comment by rr-nick-tan** [comment]
@krmahadevan , right, this is the template https://github.com/samuelmeuli/action-maven-publish/blob/master/settings.xml used by the samuelmeuli/action-maven-publish https://github.com/samuelmeuli/action-maven-publish/blob/master/index.js#L6

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -28,3 +30,15 @@ jobs:
           ./mvnw --version
           cp -f remote-test/src/test/resources/grapeConfig.xml ~/.groovy
           ./mvnw -e -DskipIntTest=false clean install jacoco:report
+
+      - name: Publish snapshots
+        uses: samuelmeuli/action-maven-publish@v1
+        # publish snapshot only when push to master branch
+        if: github.event_name == 'push'
```
Comment (juherr) [inline .github/workflows/ci.yml:37]: should only triggered if java is `8` (or `11` if we decide to drop the support of java 8 for this package)

**Comment by juherr** [comment]
Hi,
Any update about this subject? I had a look and everything looks good. I don't understand why it is not working as expected.

**Comment by krmahadevan** [comment]
> Hi,
> 
> Any update about this subject? I had a look and everything looks good. I don't understand why it is not working as expected.

I haven't been able to get to it. Currently dealing with a health crisis and yet to be discharged from hospital. 

In case you have a pair of credentials that can be used can you please try using that @juherr ?

**Comment by juherr** [comment]
@missedone It is working better but it is not yet perfect. I don't understand the auth issue with the last package only.

@krmahadevan Take care of yourself

File: .github/workflows/ci.yml (hunk 2 above)
Comment (rr-nick-tan) [reply .github/workflows/ci.yml]: good point, updated the workflow to publish on java 8 only.

**Comment by rr-nick-tan** [comment]
@krmahadevan take care

**Comment by missedone** [comment]
@juherr @krmahadevan to unblock the issue, i'm going to use the github package as the maven repository, we can think about the sync to central maven repo later.

to be able to use github package, I have to enable it on the public repo
image

let me know if you're OK.

**Comment by missedone** [comment]
i have to fix the CI and release process before fixing the issue https://github.com/testng-team/testng-remote/issues/127

**Comment by juherr** [comment]
@missedone We fixed the secret configuration last week and succeed to release TestNG with that configuration.
Please, could you have a new try with the org secrets?
image

**Comment by zanoon2020** [comment]
Dear @missedone @juherr any update on this , thanks in advance.

File: .github/workflows/ci.yml (hunk 2 above)
Comment (missedone) [reply .github/workflows/ci.yml:39]: once https://github.com/testng-team/testng-remote/pull/128 merged, we will drop the support of java 8

**Comment by missedone** [comment]
@krmahadevan @juherr , the gpg signing issue is solved, but now got 401 error when uploading artifacts to OSSRH, is it a different credentials on the maven central repo?

> Error:  Failed to execute goal org.sonatype.plugins:nexus-staging-maven-plugin:1.6.13:deploy (injected-nexus-deploy) on project testng-remote-test: Failed to deploy artifacts: Could not transfer artifact org.testng.testng-remote:testng-remote6_9_7:jar:javadoc:1.5.0-20230604.233739-1 from/to ossrh (https://s01.oss.sonatype.org/content/repositories/snapshots): Transfer failed for https://s01.oss.sonatype.org/content/repositories/snapshots/org/testng/testng-remote/testng-remote6_9_7/1.5.0-SNAPSHOT/testng-remote6_9_7-1.5.0-20230604.233739-1-javadoc.jar 401 Unauthorized -> [Help 1]

https://github.com/testng-team/testng-remote/actions/runs/5171772022/jobs/9315501084


I'm following the publishing procedure per the articles:
* https://bjansen.github.io/java/2021/02/03/deploying-to-maven-central-using-github-actions.html
* https://central.sonatype.org/publish/publish-guide/#a-complete-example-pom

**Comment by krmahadevan** [comment]
@missedone - I haven't changed anything on the Nexus credentials 

image

I think its still referring to @cbeust credentials (since we are still publishing to `org.testng` space which is owned by cedric). 

Do you remember what user id we were using to publish testng remote ?

**Comment by missedone** [comment]
@krmahadevan i have no idea whose user ID is used. If you have nexus credentials, could you enter your nexus user name and password in this repo’s [secrets ](https://github.com/testng-team/testng-remote/settings/secrets/actions) ?

**Comment by krmahadevan** [comment]
@missedone - I have added my credentials as seen below

image


But I am not sure if they work in this case, because I think the credentials are tied to the group id into which publishing is being done.
My credentials can publish into `com.rationaleemotions` but we are trying to publish into `org.testng`.

**Comment by missedone** [comment]
right, the credentials tie to the group ID

BTW, I noticed the group id `org.testng` is not available on the s01.oss snapshot repo, it might be the issue.
https://s01.oss.sonatype.org/content/repositories/snapshots/org/testng

image

**Comment by krmahadevan** [comment]
I am not sure what would be the url for the snapshot repo. But I tried `oss.sonatype` url and it says directory browsing disabled

But the search UI on sonatype shows that the snapshot group id for testng is available

image

File: pom.xml [hunk 3]
```diff
@@ -202,15 +201,102 @@
...
+                    <plugin>
+                        <groupId>org.sonatype.plugins</groupId>
+                        <artifactId>nexus-staging-maven-plugin</artifactId>
+                        <version>1.6.13</version>
+                        <extensions>true</extensions>
+                        <configuration>
+                            <serverId>ossrh</serverId>
+                            <!-- For pre 2021 legacy projects you may use https://oss.sonatype.org -->
+                            <!-- See https://central.sonatype.org/publish/release/#login-into-ossrh for details -->
+                            <nexusUrl>https://s01.oss.sonatype.org</nexusUrl>
```
Comment (krmahadevan) [inline pom.xml:283]: Can we try changing this URL to `https://oss.sonatype.org` and see if that helps ?

I use that in my projects which I publish to Nexus 

https://github.com/RationaleEmotions/SimpleSe/blob/master/pom.xml#L134

**Comment by missedone** [comment]
oh man, you saved my day, the trick worked

**Comment by missedone** [comment]
@krmahadevan @juherr , pls review the PR

i'm going to create a separate PR for the release job

**Review by krmahadevan — APPROVED** [review]

**Comment by krmahadevan** [comment]
@missedone - I will go ahead and delete those credential keys that I added to this repo, so that we dont cause any additional confusions ?

**Review by juherr — APPROVED** [review]

**Pull request merged**